theano/tensor/tests/test_opt.py

```diff
@@ -3547,6 +3554,57 @@ def test2(self):
 
         self._verify_stack_trace(f)
 
+    def test_useless_alloc_with_shape_one(self):
+        alloc_lift = out2in(local_useless_alloc)
+        x = shared(self.rng.randn(2,))
+        y = shared(self.rng.randn())
+        z = shared(self.rng.randn(1, 1))
+        w = shared(self.rng.randn(1, 1))
+        alloc_x = tensor.alloc(x, 1, 3, 2)
+        alloc_y = tensor.alloc(y, 1, 1)
+        alloc_z = tensor.alloc(z, 1, 1, 2)
+        alloc_w = tensor.alloc(w, 1, 2)
+
+        g = FunctionGraph([x, y, z, w], [alloc_x, alloc_y, alloc_z, alloc_w])
+        self.assertTrue(str(g) == ("[Alloc(<TensorType(float64, vector)>, "
+                                   "TensorConstant{1}, "
+                                   "TensorConstant{3}, "
+                                   "TensorConstant{2}), "
+
+                                   "Alloc(<TensorType(float64, scalar)>, "
+                                   "TensorConstant{1}, "
+                                   "TensorConstant{1}), "
+
+                                   "Alloc(<TensorType(float64, matrix)>, "
+                                   "TensorConstant{1}, "
+                                   "TensorConstant{1}, "
+                                   "TensorConstant{2}), "
+
+                                   "Alloc(<TensorType(float64, matrix)>, "
+                                   "TensorConstant{1}, "
+                                   "TensorConstant{2})]"))
+
+        alloc_lift.optimize(g)
+        self.assertTrue(str(g) == "[DimShuffle{x,0,1}"
+                                  "(Alloc(<TensorType(float64, vector)>, "
+                                  "TensorConstant{3}, "
+                                  "TensorConstant{2})), "
+
+                                  "DimShuffle{x,x}"
+                                  "(<TensorType(float64, scalar)>), "
+
+                                  "DimShuffle{x,0,1}"
+                                  "(Alloc(<TensorType(float64, matrix)>, "
```

Comment:
Because it makes it possible to combine the dimshuffle with another one, or lift it further.
These optimizations were identified to help simplify the case in https://github.com/Theano/Theano/issues/2943#issuecomment-184453452, where operations like, alloc, dimshuffle, and reshape do more than one thing, which can all cancel out if they are properly separated.

It may be possible to add optimizations combining these operations again after the simplification is done, after the 'specialize' phase or so, but that is another issue.